### PR TITLE
Various micro optimizations of the component boot flow.

### PR DESCRIPTION
**Pull request by cramforce**
- Removes unnecessary `hasAttributes` call from `propagateAttributes`
- Bundles several `classList.add` operations into one.
- Add special case for AMP Doc retrieval that is faster if no shadow doc was ever added which is common.
- Makes hot `getTopWindow` function faster by ensuring the property that is rarely defined is at least known.

### CODE REVIEW

File: extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js [hunk 1]
```diff
@@ -167,20 +166,50 @@ export class AmpFlyingCarpet extends AMP.BaseElement {
       this.element,
       this.children_
     );
-    listen(this.element, AmpEvents.BUILT, this.layoutBuiltChild_.bind(this));
+    this.observeNewChildren_();
```
Comment (jridgewell) [inline extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js]: Did you mean to remove these changes? It seems to be the same changes as https://github.com/ampproject/amphtml/pull/23774.

I'm skipping this file assuming that is true.

File: src/base-element.js [hunk 2]
```diff
@@ -678,8 +678,9 @@ export class BaseElement {
     attributes = isArray(attributes) ? attributes : [attributes];
     for (let i = 0; i < attributes.length; i++) {
       const attr = attributes[i];
-      if (this.element.hasAttribute(attr)) {
-        element.setAttribute(attr, this.element.getAttribute(attr));
+      const val = this.element.getAttribute(attr);
+      if (val !== null) {
```
Comment (jridgewell) [inline src/base-element.js]: Nit: flip.

File: src/custom-element.js [hunk 3]
```diff
@@ -803,9 +803,11 @@ function createBaseCustomElementClass(win) {
       this.isConnected_ = true;
 
       if (!this.everAttached) {
-        this.classList.add('i-amphtml-element');
-        this.classList.add('i-amphtml-notbuilt');
-        this.classList.add('amp-notbuilt');
+        this.classList.add(
```
Comment (jridgewell) [inline src/custom-element.js]: This breaks IE11 support: https://developer.mozilla.org/en-US/docs/Web/API/Element/classList#Browser_compatibility. We could use `this.className += ' foo bar baz'`, but that'll cause a tokenization of all current classnames, too.

File: src/custom-element.js [hunk 4]
```diff
@@ -862,8 +864,7 @@ function createBaseCustomElementClass(win) {
           this.tryUpgrade_();
         }
         if (!this.isUpgraded()) {
-          this.classList.add('amp-unresolved');
-          this.classList.add('i-amphtml-unresolved');
+          this.classList.add('amp-unresolved', 'i-amphtml-unresolved');
```
Comment (jridgewell) [inline src/custom-element.js]: Ditto.

File: src/service.js [hunk 5]
```diff
@@ -318,7 +318,16 @@ export function getParentWindow(win) {
...
-  return win.__AMP_TOP || win;
+  const top = win.__AMP_TOP;
+  if (top === null) {
+    return win;
+  }
+  if (top != null) {
+    return top;
+  }
+  // Make this defined but null to speed up future lookups.
+  win.__AMP_TOP = null;
```
Comment (jridgewell) [inline src/service.js]: We could set this to `win` instead? Then keep the same `||`: `win.__AMP_TOP || (win.__AMP_TOP = win)`

File: extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js (hunk 1 above)
Comment (cramforce) [reply extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js]: Yeah, just forgot to the update

File: src/custom-element.js (hunk 3 above)
Comment (kristoferbaxter) [reply src/custom-element.js]: We could introduce a transform for this so the module build can use `TokenList.add(...params)` but transpile to separate `TokenList.add(param)` for the legacy build.

Comment (cramforce) [reply src/custom-element.js]: I'll polyfill instead.

**Comment by cramforce** [comment]
PTAL

File: src/custom-element.js (hunk 3 above)
Comment (kristoferbaxter) [reply src/custom-element.js]: Let's make sure the polyfill is not added to the module build too.

**Pull request merged**